### ISSUE TITLE
Publish a BOM (Bill of Materials)

### ISSUE THREAD

**Issue by ansman** [google/dagger#2122]
Now that Hilt is a thing it would be helpful if Dagger published a BOM (Bill of Materials) along with the main artifacts.

There are a couple of reasons why this would be nice:
- With Hilt you are depending on AndroidX Hilt as well which might be pulling in a different version of one of the artifacts
- There are a lot of versions to sync

**Comment by bcorso** [comment]
Hi @ansman, are you referring to the [Maven BOM POM](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#bill-of-materials-bom-poms)?

Sorry, it's my first time looking into this and I'm still trying to figure out how this would be used in practice. Is the idea that users would use a [Bom import](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import) to avoid specifying the version on individual artifacts? For example:

```groovy
dependencies {
    // Import the Dagger BOM
    implementation platform('com.google.dagger:bom:2.29.1')

    // Define Dagger dependencies without versions
    implementation 'com.google.dagger:dagger'
    annotationProcessor 'com.google.dagger:dagger-compiler'
}
```

Is there some other benefits that I'm not covering with the above?

If the above understanding is correct, the main question I have is: does the BOM provide any additional benefits over handling this within your build system. For example, I usually see this use case handled in Gradle by defining an extension property for the Dagger version:

```groovy
// project build.gradle
buildscript {
    ext.daggerVersion = "2.29.1"
}

// app/library build.gradle
dependencies {
    implementation "com.google.dagger:dagger:$daggerVersion"
    annotationProcessor "com.google.dagger:dagger-compiler:$daggerVersion"
}
```

Which seems to provide similar guarantees to the BOM, but it can be handled consistently across all libraries whether or not the library defines a BOM.

**Comment by ansman** [comment]
Yes, that is what I'm referring to.

One benefit of a BOM is, like you mention, to avoid having to specify the version for dependencies belonging to the same BOM.

There is however another, much more important, benefit namely transitive dependencies.
Say that dagger has 3 artifacts named `api`, `compiler`, `utils` as an example.
You include the api and compiler using the same version variable but you are not using the `utils` artifact so that one is omitted.

Then you include another third party library that is using `utils` but of a different version. Without the BOM the `api` and `compiler` artifacts gets resolved using the normal resolution strategy (likely the highest version wins) but the `utils` gets included using whatever version the other third party library uses.

This example is what prompted me to create this issue. I was not depending on all artifacts that dagger has directly. But the AndroidX hilt artifacts were pulling some of the ones I weren't which led to issues because not all dagger artifacts were of the same version.

This behavior can perhaps be achieved manually using Gradle but it would likely force you to keep the list of Dagger artifacts up to date in your project which doesn't seem likely.

You can read more about Gradle's BOM support here: https://docs.gradle.org/current/userguide/platforms.html

Lastly I would like to mention that implementing support for a BOM is likely very small task and I might even be able to contribute the change via a PR.

**Comment by bcorso** [comment]
Thanks @ansman, that makes sense. I agree that it would be nice to support this. 

However, it'll likely be more than just a small task from our side. Dagger is built with Bazel and we have quite a bit of custom infrastructure that converts those build targets into maven artifacts with the correctly generated pom files. We'll have to take a look and see what the best way to integrate the BOM into our current setup.

**Comment by ansman** [comment]
Right, that makes sense. If it's any consolation there is no Gradle plugin for this either as far as I could find. But writing your own is fairly straight forward as long as you can list all the artifacts and their version. Here is our implementation which is for Gradle:
```kotlin
private val dependencies: Provider<List<Dependency>> = project.provider {
    project.rootProject.allProjects.flatMap { projects ->
        projects
            .asSequence()
            .filter { it.extensions.findByType(PublishingExtension::class.java) != null }
            .flatMap { 
                it.publishing.publications
                    .withType(MavenPublication::class.java)
                    .asSequence()
                    .map { publication ->
                        Dependency(
                            groupId = publication.groupId,
                            artifactId = publication.artifactId,
                            version = publication.version,
                        )
                }
            }
            .toList()
    }
}

val bom = project.publishing.publications.register("bom", MavenPublication::class.java)
bom.configure {
    it.artifactId = artifactId.get()
    with(it.pom) {
        packaging = "pom"
        withXml { xml ->
            with(xml.asNode().appendNode("dependencyManagement").appendNode("dependencies")) {
                for (dependency in dependencies.get()) {
                    append(dependency)
                }
            }
        }
    }
}

private fun Node.append(dependency: Dependency) = with(appendNode("dependency")) {
    appendNode("groupId", dependency.groupId)
    appendNode("artifactId", dependency.artifactId)
    appendNode("version", dependency.version)
    if (dependency.type != null) appendNode("type", dependency.type)
    if (dependency.scope != null) appendNode("scope", dependency.scope)
}

data class Dependency(
    val groupId: String,
    val artifactId: String,
    val version: String,
    val type: String? = null,
    val scope: String? = null,
)
```

**Comment by ernestp** [comment]
Firebase SDK already using BOM. Perhaps it will be easy to collaborate with Firebase team on implementing BOM for Dagger. (https://firebase.google.com/docs/android/learn-more?authuser=0#bom)
